domains/client/domain-executor/src/domain_block_processor.rs

```diff
@@ -208,20 +219,11 @@ where
         extrinsics: Vec<Block::Extrinsic>,
         digests: Digest,
     ) -> Result<DomainBlockResult<Block, PBlock>, sp_blockchain::Error> {
-        let primary_number = to_number_primitive(primary_number);
-
-        if to_number_primitive(parent_number) + 1 != primary_number {
-            return Err(sp_blockchain::Error::Application(Box::from(format!(
-                "Wrong domain parent block #{parent_number},{parent_hash} for \
-                primary block #{primary_number},{primary_hash}, the number of new \
-                domain block must match the number of corresponding primary block."
-            ))));
-        }
-
         // Although the domain block intuitively ought to use the same fork choice
         // from the corresponding primary block, it's fine to forcibly always use
-        // the longest chain for simplicity as we manually build all the domain
-        // branches by literally following the primary chain branches anyway.
+        // the longest chain for simplicity as we do not know whether or not the
+        // domain block is the new best block here, and we will manually reset the
+        // new best domain block to the correct one after processing the primary blocks
```

Comment:
The reasoning is not super clear. The reason for using the longest chain is still to follow the primary chain branches, the reason for the new best block handling is caused by the change of primary block and domain block is not mapped in 1:1.

```suggestion
        // the longest chain for simplicity as we manually build the domain branches
        // by following the primary chain branches. Due to the possibility of domain
        // branch transitioning to a lower fork caused by the change that a primary block
        // can possibility produce no domain block, it's important to note that now we
        // need to ensure the domain block built from the latest primary block is the
        // new best domain block after processing each imported primary block.
```

Reply:
> The reason for using the longest chain is still to follow the primary chain branches

This is not correct, since the primary block and domain block are not mapped in 1:1, using the longest chain may cause the best fork of the domain chain set to the wrong one temporarily and recorrect after we re-import the best domain block.

The reason for using the longest chain is that we do not know if the domain block is the new best block or not here, but your proposal do makes it clearer and is applied.

Reply:
>The reason for using the longest chain is that we do not know if the domain block is the new best block or not here

Not true. The reason is not that we do not know if the domain block is the new best block or not here, you don't have to know that to build the domain branches as I said in another comment: conceptually if you re-apply the fork choice of every consensus block on the corresponding domain block, the best domain block would be constructed naturally and correctly.